Docker images

matyasselmeci opened the pull request:
This commit adds Dockerfiles and supporting files for creating htcondor images. The execute node has received the most attention, but there are also images for the cm, submit node, and minicondor, mostly for testing. See README.md for documentation.

GregThain commented on src/condor_contrib/docker/README.md:
```diff
@@ -0,0 +1,142 @@
+The images for the various roles (`execute`, `cm`, `mini`, `submit`)
+are built on top of the `htcondor/base` image, which gets built first.
+`htcondor/base` contains the HTCondor software, some common configuration
+and scripts.
```
Just to be clear, we should mention that one need to be root or in the docker group to run this makefile.

GregThain commented on src/condor_contrib/docker/execute/packagelist.txt:
```diff
@@ -0,0 +1,44 @@
+# GNU's bc (a numeric processing language) and dc (a calculator):
+bc
```
Do I want to know what requires bc?

matyasselmeci replied:
Shell scripts that don't know how to use `$(( ))` or `expr`?

I built the package list by getting a list of what gets installed by `yum groupinstall Base` and removing what's obviously unnecessary.

GregThain commented on src/condor_contrib/docker/README.md:
```diff
@@ -140,3 +140,6 @@ Known Issues
 
+- cgroups support is not yet implemented.
+
+- Docker universe support is not yet implemented.
```
Other things that probably don't work: Singularity

matyasselmeci replied:
I think if I installed Singularity into the image, it might work. Some of the worker nodes made by OSG folks have Singularity installed in them but I don't know if they're run privileged or not.

GregThain commented on src/condor_contrib/docker/Makefile:
```diff
@@ -0,0 +1,102 @@
+REGISTRY:=
+DATE:=$(shell date +%Y%m%d)
+VERSION:=8.9.5
```
It's a bit more quoting headache, but we could rename this to Makefile.in, make it a cmake configure_file, and have cmake expand the VERSION automatically.

GregThain commented on src/condor_contrib/docker/README.md:
```diff
@@ -0,0 +1,145 @@
+
+The execute node container can connect to an existing HTCondor pool
+via token authentication (recommended) or pool
+password.  The container uses CCB so exposing ports is not necessary.
```
I would further clarify this by saying "because of CCB, this container only needs outbound connectivity to the central manager and submit machines even to run jobs and condor_ssh_to_job.

GregThain commented on src/condor_contrib/docker/README.md:
````diff
@@ -0,0 +1,145 @@
+
+Start the container:
+```console
+dockerhost$ docker run --env-file=~/condorexec/env \
````
We may want to add additional docker run best practice options here, like:
--rm=true
--name="htcondor-exec-node"
--memory=4400 (or some epsilon over MEMORY)

GregThain commented on src/condor_contrib/docker/README.md:
```diff
@@ -0,0 +1,145 @@
+### Example
+
+This example is for creating an execute node and adding it to a pool using
+token auth.  The execute node will have 2 CPUs and 4 GB of RAM, and will
```
2 cpus in static slots, right?

matyasselmeci replied:
No, pslots.

matyasselmeci replied:
Yes.

GregThain commented on src/condor_contrib/docker/README.md:
```diff
@@ -0,0 +1,145 @@
+`make`.
+
+See `make help` for additional targets and options.
+
```
Should there also be a section on customizing images, e.g. how to add your own software to an execute node?

matyasselmeci commented:
I think I addressed all your comments, Greg. Want to take another look? I didn't add `--rm` to docker run but I did add `--detach` instead since there's not much point to having a server eating up a terminal.
